### PR TITLE
Modify github-to-sops directory structure for publishing to PyPI

### PR DESCRIPTION
**Pull request by rjwignar**
Summary
---
This Pull Request modifies the directory structure of `github-to-sops` to one that can be packaged and published to PyPI.
This was done using Simon Willison's [cookiecutter template](https://github.com/simonw/python-lib-template-repository) to create a mock project and replicating mock project's directory structure in this branch.

These changes should allow us to ship `github-to-sops` to ChatCraft via `pip`, as per https://github.com/tarasglek/chatcraft.org/issues/324
## The cookiecutter template

The cookiecutter template:
- creates a `pyproject.toml` file configured with `setuptools`
  - includes `pytest` as a test dependency
- adds a README
- creates `package_name/__init__.py`
- creates `tests/test_package_name.py`
- Adds two workflows in `.github/workflows`:
  - one to run tests against Python 3.8 to 3.12
  - one to publish package releases to PyPI.

## My modifications
- added `pyproject.toml` based on what is produced by the template. Content modified to be specific to github-to-sops
- Kept original `README.md`
- moved `github-to-sops` to `github-to-sops/__init__.py`
- added `tests/test_github_to_sops` with empty tests
- copied `.github/workflows` to directory (as the workflow content doesn't depend on the project name)

## Next Steps
Assuming this directory structure is correct, the next steps would be to use PyPI's [Trusted Publishing](https://docs.pypi.org/trusted-publishers/) mechanism to publish github-to-sops to PyPI.

Below is an example of the Trusted Publishing form Willison filled to publish his own library, [datasette-test](https://github.com/datasette/datasette-test)
image

We would have to fill in this form with this repo's info.
I might be able to fill in the form myself but I wanted to first confirm these changes with @tarasglek.

### CODE REVIEW

**Comment by rjwignar** [comment]
I was a little iffy on the `tests` folder. Should I remove it and the reference to `pytest` in `pyproject.toml`?
Removing tests would also make `.github/workflows/test.yml` unnecessary.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+[project]
+name = "github-to-sops"
+version = "0.1"
+description = ""
```
Comment (humphd) [inline pyproject.toml]: Let's add this

**Review by humphd — COMMENTED** [review]
@rjwignar great work!  Thanks for pushing on this for 1.1.

I'm not a Python dev, but I wonder if this is right `github-to-sops → github-to-sops/__init__.py`. I'd be surprised to find the source for a module in `__init__.py`.  Maybe @tarasglek knows.

How do secrets work for the CI publish workflow?  I assume @tarasglek has to do this from his account somehow?

File: tests/test_github-to-sops.py
Comment (WangGithub0) [inline tests/test_github-to-sops.py:1]: I am uncertain whether we should add the most basic test？

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      with:
+        python-version: ${{ matrix.python-version }}
+        cache: pip
+        cache-dependency-path: pyproject.toml
+    - name: Install dependencies
+      run: |
+        pip install '.[test]'
+    - name: Run tests
+      run: |
+        pytest
```
Comment (WangGithub0) [inline .github/workflows/publish.yml]: why this step is different from the `Run tests` in `test.yml`?

File: tests/test_github-to-sops.py
Comment (rjwignar) [reply tests/test_github-to-sops.py:1]: I agree. I think I will remove this file.
Depending on if we want tests at all, I might have to remove or modify some other files that reference pytest
(e.g. `.github/workflows/test.yml`).

**Comment by rjwignar** [comment]
@kliu57 could you take a look at `github-to-sops/__init__.py`?
I moved the python library from the file `github-to-sops` to `github-to-sops/__init__.py`.

I've seen some projects with empty an `__init__.py`, and some with `__init__.py` containing programs.

Does it make sense to have the `github-to-sops` program entirely inside `__init__.py`?

**Comment by humphd** [comment]
@tarasglek can you add @rjwignar to the collaborators for this repo?

File: .github/workflows/publish.yml (hunk 2 above)
Comment (rjwignar) [reply .github/workflows/publish.yml]: I'm not sure why. Both workflows were generated by https://github.com/simonw/python-lib-template-repository.
However, we can remove this step entirely if we don't include pytest in this library.

**Comment by kliu57** [comment]
> @kliu57 could you take a look at `github-to-sops/__init__.py`? I moved the python library from the file `github-to-sops` to `github-to-sops/__init__.py`.
> 
> I've seen some projects with empty an `__init__.py`, and some with `__init__.py` containing programs.
> 
> Does it make sense to have the `github-to-sops` program entirely inside `__init__.py`?

Back when I published something to PyPI last term I used an empty `__init__.py `file. [See blog](https://dev.to/katiel/software-release-using-pypi-1ddi) to see what I did and what instructional video I followed. Did you read some instructions from somewhere which told you to put the program inside `__init__.py`?

**Comment by rjwignar** [comment]
> > @kliu57 could you take a look at `github-to-sops/__init__.py`? I moved the python library from the file `github-to-sops` to `github-to-sops/__init__.py`.
> > I've seen some projects with empty an `__init__.py`, and some with `__init__.py` containing programs.
> > Does it make sense to have the `github-to-sops` program entirely inside `__init__.py`?
> 
> Back when I published something to PyPI last term I used an empty `__init__.py `file. [See blog](https://dev.to/katiel/software-release-using-pypi-1ddi) to see what I did and what instructional video I followed. Did you read some instructions from somewhere which told you to put the program inside `__init__.py`?

The directory structure generated by https://github.com/simonw/python-lib-template-repository includes program code inside `__init__.py`. I moved the `github-to-sops` code to `__init__.py` based on that.

Here's the `__init__.py` generated by the template:
https://github.com/rjwignar/test/blob/main/test/__init__.py

**Comment by kliu57** [comment]
> > > @kliu57 could you take a look at `github-to-sops/__init__.py`? I moved the python library from the file `github-to-sops` to `github-to-sops/__init__.py`.
> > > I've seen some projects with empty an `__init__.py`, and some with `__init__.py` containing programs.
> > > Does it make sense to have the `github-to-sops` program entirely inside `__init__.py`?
> > 
> > 
> > Back when I published something to PyPI last term I used an empty `__init__.py `file. [See blog](https://dev.to/katiel/software-release-using-pypi-1ddi) to see what I did and what instructional video I followed. Did you read some instructions from somewhere which told you to put the program inside `__init__.py`?
> 
> The directory structure generated by https://github.com/simonw/python-lib-template-repository includes program code inside `__init__.py`. I moved the `github-to-sops` code to `__init__.py` based on that.
> 
> Here's the `__init__.py` generated by the template: https://github.com/rjwignar/test/blob/main/test/__init__.py

I see, so it was automatically generated by the tool. I've never used this tool, I did my project restructuring manually.

**Comment by humphd** [comment]
If Simon puts the code in `__init__.py`, it must be the right way to do it--he publishes tons of pypi stuff.  Let's not overthink it.

File: pyproject.toml (hunk 1 above)
Comment (rjwignar) [reply pyproject.toml]: Good catch. Description added as of b962ad31b3091c43e4c22a0bd4fc83262debc776:
image

File: tests/test_github-to-sops.py
Comment (rjwignar) [reply tests/test_github-to-sops.py:1]: Removed `tests/test_github-to-sops` as of 278bda4fe085d0a0ef0ef6b756a3a7f02aeb09c5, leaving the `tests` folder empty.
The test.yml workflow still passes without errors: https://github.com/rjwignar/github-to-sops/actions/runs/7735949309

@tarasglek, Simon's template configures the library with pytest as an optional dependency.
We currently don't have any tests setup for github-to-sops, but I see unit tests are something you'd like to see added to `github-to-sops`:
image

With that in mind, would you like this PR to be configured with pytest setup so tests can be added in the future?
i.e. keep the following in the PR?:
- an empty `test` folder
- `.github/workflows/test.yml`, 
- the `test` jobs inside `publish.yml`, 
- `pytest` as an optional dependency in `pyproject.toml`)

**Comment by rjwignar** [comment]
I've been able to fill out the form myself and add github-to-sops as a **pending project** on PyPI (we've **reserved** the package name github-to-sops)
> How do secrets work for the CI publish workflow? I assume @tarasglek has to do this from his account somehow?

@humphd It looks like no secrets are involved in the publish workflow.

From [Simon's blog on Trusted Publishing](https://til.simonwillison.net/pypi/pypi-releases-from-github), It looks like Trusted Publishing involves three steps:
- [x] Tell PyPI which GitHub repository should be allowed to publish a package with a specific name
  - Completed by filling out the Pending Publisher form:
  image
  - note that **release** is the configured publishing environment name
- [x] Configure GitHub Actions `publish` workflow to use the `pypa/gh-action-pypi-publish@release/v1 action`
  - `.github/workflows/publish.yml` generated by [Simon's template](https://github.com/simonw/python-lib-template-repository)
  - note that **release** is configured as the environment name in the `deploy` job of `publish.yml`, to match the Pending Publisher setup:
  https://github.com/tarasglek/github-to-sops/blob/278bda4fe085d0a0ef0ef6b756a3a7f02aeb09c5/.github/workflows/publish.yml#L30-L35
- [ ] Publish a release to GitHub that triggers the workflow
  - Publishing releases is restricted to Taras unless anyone else is granted **Write** access to the repo.

**Comment by rjwignar** [comment]
> If Simon puts the code in `__init__.py`, it must be the right way to do it--he publishes tons of pypi stuff. Let's not overthink it.

I've taken a look at some of his recent projects and he keeps source code into submodules that are imported by `__init__.py`.
**However**, keeping all the code in `__init__.py` is possible too. [pylons/colander](https://github.com/Pylons/colander) has its core code (~3000 lines) inside [\_\_init\_\_.py](https://github.com/Pylons/colander/blob/main/src/colander/__init__.py)!

I think for this PR, we should keep `__init__.py` as is, then we can look into moving the code into submodules in a follow-up issue.

**Comment by tarasglek** [comment]
Invited you guys to the repo

**Comment by tarasglek** [comment]
@rjwignar I don't think you need anything from me to proceed, please push a release.

**Pull request merged**